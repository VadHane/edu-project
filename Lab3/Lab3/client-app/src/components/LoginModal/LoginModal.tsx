import {
    Alert,
    Button,
    Modal,
    TextField,
    Typography,
    CircularProgress,
    Snackbar,
} from '@mui/material';
import { Box } from '@mui/system';
import React, { FunctionComponent, useEffect, useState } from 'react';
import { useNavigate } from 'react-router-dom';
import {
    INCORRECT_EMAIL_EXCEPTION,
    LENGTH_OF_PASSWORD_EXCEPTION,
} from '../../exceptions';
import { useTypedSelector } from '../../hooks/useTypedSelector';
import { useAuthActions } from './../../hooks/useAuthActions';
import './LoginModal.css';

const LoginModal: FunctionComponent = () => {
    const navigate = useNavigate();

    const [email, setEmail] = useState<string>('');
    const [password, setPassword] = useState<string>('');

    const [emailError, setEmailError] = useState<string>('');
    const [passwordError, setPasswordError] = useState<string>('');
    const [errorMessage, setErrorMessage] = useState<string>('');

    const { authorization } = useAuthActions();
    const { isAuth, error, isLoading } = useTypedSelector((state) => state.auth);

    useEffect(() => {
        if (isAuth) {
            navigate('/');
        }
    }, [isAuth]);

    useEffect(() => {
        setErrorMessage('');
        setEmailError('');
        setPasswordError('');
    }, [email, password]);

    useEffect(() => {
        if (error !== null) {
            setErrorMessage(error);
        }
    }, [error]);

    const validation = (): boolean => {
        const pattern = /^[^ ]+@[^ ]+\.[a-z]{2,3}$/;
        if (!email.match(pattern)) {
            setEmailError(INCORRECT_EMAIL_EXCEPTION);
            return false;
        }

        if (password.trim().length < 3) {
            setPasswordError(LENGTH_OF_PASSWORD_EXCEPTION);
            return false;
        }

        return true;
    };

    const onClose = () => {
        navigate('/');
    };

    const onSendForm = () => {
        if (!validation()) {
            return;
        }

        authorization(email, password);
    };

    const loginModalNode: React.ReactNode = (
        <>
            <Modal
                open={true}
                onClose={onClose}
                aria-labelledby="modal-modal-title"
                aria-describedby="modal-modal-description"
            >
                <Box className="modal-window">
                    {isLoading ? (
                        <Box className="loading-circle">
                            <CircularProgress />
                        </Box>
                    ) : (
                        <>
                            <Typography
                                id="modal-modal-title"
                                variant="h4"
                                component="h2"
                            >
                                Login
                            </Typography>

                            <TextField
                                id="standard-basic"
                                className="text_field"
                                label="Email"
                                variant="standard"
                                onChange={(e) => setEmail(e.target?.value)}
                                value={email}
                                error={emailError !== ''}
                                helperText={emailError}
                            />
                            <TextField
                                className="text_field"
                                variant="standard"
                                label="Password"
                                type="password"
                                autoComplete="current-password"
                                onChange={(e) => setPassword(e.target?.value)}
                                value={password}
                                error={passwordError !== ''}
                                helperText={passwordError}
                            />

                            <Button
                                className="window_button"
                                variant="contained"
                                onClick={onSendForm}
                            >
                                Send
                            </Button>
                        </>
                    )}
                </Box>
            </Modal>
        </>
    );

    const statusNode: React.ReactNode = (
        <>
            <Snackbar
                anchorOrigin={{ vertical: 'top', horizontal: 'center' }}
                open={errorMessage !== ''}
                key={'top' + 'center'}
            >
                <Alert severity="error" sx={{ width: '100%' }}>
                    {error}
                </Alert>
            </Snackbar>
        </>
    );

    return (
        <div>
            {loginModalNode}
            {statusNode}
        </div>
    );
};

export default LoginModal;
