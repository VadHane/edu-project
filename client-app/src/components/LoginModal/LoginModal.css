.modal-window {
    position: absolute;
    top: 50%;
    left: 50%;
    transform: translate(-50%, -50%);
    width: 400px;
    background-color: #fff;
    border: 2px solid #000;
    box-shadow: 24px;
    padding: 40px;
    display: flex;
    flex-direction: column;
    text-align: center;
}

.modal-window .text_field {
    margin-top: 20px;
}

.modal-window .window_button {
    width: 50%;
    margin: 0 auto;
    margin-top: 70px;
}

.modal-window .loading-circle {
    margin: 0 auto;
    display: block;
}
