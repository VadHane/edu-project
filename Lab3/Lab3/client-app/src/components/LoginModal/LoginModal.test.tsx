import { render } from '@testing-library/react';
import React from 'react';
import LoginModal from './LoginModal';

jest.mock('../../hooks/useTypedSelector', () => ({
    ...jest.requireActual('../../hooks/useTypedSelector'),
    useTypedSelector: () => ({ isAuth: false, error: null, isLoading: false }),
}));

jest.mock('./../../hooks/useAuthActions', () => ({
    ...jest.requireActual('./../../hooks/useAuthActions'),
    useAuthActions: () => ({ authorization: jest.fn() }),
}));

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: jest.fn(),
}));

describe('Test modal window for login.', () => {
    test('Snapshot', () => {
        const utils = render(<LoginModal />);

        expect(utils).toMatchSnapshot();
    });
});
