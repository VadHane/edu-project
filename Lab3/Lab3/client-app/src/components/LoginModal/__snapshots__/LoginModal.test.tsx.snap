// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test modal window for login. Snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body
    style="padding-right: 1024px; overflow: hidden;"
  >
    <div
      aria-hidden="true"
    >
      <div />
    </div>
    <div
      aria-describedby="modal-modal-description"
      aria-labelledby="modal-modal-title"
      class="MuiModal-root css-79ws1d-MuiModal-root"
      role="presentation"
    >
      <div
        aria-hidden="true"
        class="MuiBackdrop-root css-i9fmh8-MuiBackdrop-root-MuiModal-backdrop"
        style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
      />
      <div
        data-test="sentinelStart"
        tabindex="0"
      />
      <div
        class="modal-window MuiBox-root css-0"
        tabindex="-1"
      >
        <h2
          class="MuiTypography-root MuiTypography-h4 css-5lbw0b-MuiTypography-root"
          id="modal-modal-title"
        >
          Login
        </h2>
        <div
          class="MuiFormControl-root MuiTextField-root text_field css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-standard MuiFormLabel-root MuiFormLabel-colorPrimary css-aqpgxn-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for="standard-basic"
            id="standard-basic-label"
          >
            Email
          </label>
          <div
            class="MuiInput-root MuiInput-underline MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-1ptx2yq-MuiInputBase-root-MuiInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiInput-input MuiInputBase-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
              id="standard-basic"
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          class="MuiFormControl-root MuiTextField-root text_field css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-standard MuiFormLabel-root MuiFormLabel-colorPrimary css-aqpgxn-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for=":r1:"
            id=":r1:-label"
          >
            Password
          </label>
          <div
            class="MuiInput-root MuiInput-underline MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-1ptx2yq-MuiInputBase-root-MuiInput-root"
          >
            <input
              aria-invalid="false"
              autocomplete="current-password"
              class="MuiInput-input MuiInputBase-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
              id=":r1:"
              type="password"
              value=""
            />
          </div>
        </div>
        <button
          class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root window_button css-sghohy-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          Send
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
      <div
        data-test="sentinelEnd"
        tabindex="0"
      />
    </div>
  </body>,
  "container": <div
    aria-hidden="true"
  >
    <div />
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
